### ISSUE TITLE
Questions on multi-gpu inference performance

### ISSUE THREAD

**Issue by Rogerwyf** [huggingface/text-generation-inference#1129]
Hi there, first of all thank you guys for the work on TGI and it has been amazing! This is not a bug report but rather some questions I have on mulit-gpu inference performance with TGI. I have done some benchmarking with TGI `v1.1.0` on EKS on `llama2-7b-chat-hf` and `llama2-13b-chat-hf` with A10G (g5.12xlarge) and had an interesting observation that **sharding the model over more GPUs reduces the token-level latency**

Below are some numbers for`llama2-7b-chat-hf` with `fp16` from my benchmark. All tokens in the input prompts are fixed-length and generated randomly from the model's tokenizer. For each configuration I ran the inference **non-concurrently** for 100 prompts. 

| \# GPU | Input Length | max_new_token | Average Token Latency (ms / token) | Median Token Latency - (ms / token) | P99 Token Latency (ms / token) |
| ------ | ------------ | ------------- | -------------------------------------------------------------------------- | ------------------------------------------------------------------------- | ---------------------------------------------------------------------- |
| 2      | 256          | 16            | 23.05                                                                      | 21.68                                                                     | 36.45                                                                  |
| 2      | 256          | 32            | 20.86                                                                      | 20.06                                                                     | 31.96                                                                  |
| 2      | 256          | 64            | 21.97                                                                      | 19.71                                                                     | 44.50                                                                  |
| 2      | 256          | 128           | 20.71                                                                      | 19.41                                                                     | 32.08                                                                  |
| 2      | 256          | 256           | 20.92                                                                      | 19.45                                                                     | 35.49                                                                  |
| 2      | 256          | 512           | 21.42                                                                      | 19.74                                                                     | 44.97                                                                  |
| 2      | 512          | 16            | 27.21                                                                      | 25.92                                                                     | 43.34                                                                  |
| 2      | 512          | 32            | 23.80                                                                      | 22.47                                                                     | 38.85                                                                  |
| 2      | 512          | 64            | 22.00                                                                      | 21.08                                                                     | 35.30                                                                  |
| 2      | 512          | 128           | 21.27                                                                      | 20.51                                                                     | 32.85                                                                  |
| 2      | 512          | 256           | 21.00                                                                      | 20.31                                                                     | 35.02                                                                  |
| 2      | 512          | 512           | 21.03                                                                      | 20.41                                                                     | 34.02                                                                  |
| 2      | 1024         | 16            | 37.25                                                                      | 33.85                                                                     | 59.58                                                                  |
| 2      | 1024         | 32            | 30.13                                                                      | 27.78                                                                     | 49.90                                                                  |
| 2      | 1024         | 64            | 27.24                                                                      | 24.55                                                                     | 53.29                                                                  |
| 2      | 1024         | 128           | 25.45                                                                      | 22.64                                                                     | 54.15                                                                  |
| 2      | 1024         | 256           | 24.97                                                                      | 21.92                                                                     | 53.73                                                                  |
| 2      | 1024         | 512           | 25.06                                                                      | 21.79                                                                     | 55.29                                                                  |
| 2      | 2048         | 16            | 55.77                                                                      | 50.10                                                                     | 116.37                                                                 |
| 2      | 2048         | 32            | 44.82                                                                      | 37.00                                                                     | 127.93                                                                 |
| 2      | 2048         | 64            | 34.78                                                                      | 30.14                                                                     | 84.25                                                                  |
| 2      | 2048         | 128           | 31.76                                                                      | 26.78                                                                     | 84.41                                                                  |
| 2      | 2048         | 256           | 30.33                                                                      | 25.12                                                                     | 95.78                                                                  |
| 2      | 2048         | 512           | 29.61                                                                      | 24.49                                                                     | 86.97                                                                  |
| 4      | 256          | 16            | 19.81                                                                      | 18.66                                                                     | 28.17                                                                  |
| 4      | 256          | 32            | 17.69                                                                      | 16.89                                                                     | 29.09                                                                  |
| 4      | 256          | 64            | 17.62                                                                      | 16.41                                                                     | 29.20                                                                  |
| 4      | 256          | 128           | 17.44                                                                      | 16.25                                                                     | 30.62                                                                  |
| 4      | 256          | 256           | 17.64                                                                      | 16.25                                                                     | 39.72                                                                  |
| 4      | 256          | 512           | 17.43                                                                      | 16.25                                                                     | 31.03                                                                  |
| 4      | 512          | 16            | 22.69                                                                      | 21.30                                                                     | 38.25                                                                  |
| 4      | 512          | 32            | 19.39                                                                      | 18.50                                                                     | 30.02                                                                  |
| 4      | 512          | 64            | 17.98                                                                      | 17.23                                                                     | 29.20                                                                  |
| 4      | 512          | 128           | 17.92                                                                      | 16.70                                                                     | 40.55                                                                  |
| 4      | 512          | 256           | 17.09                                                                      | 16.49                                                                     | 28.89                                                                  |
| 4      | 512          | 512           | 16.90                                                                      | 16.30                                                                     | 29.32                                                                  |
| 4      | 1024         | 16            | 36.01                                                                      | 28.74                                                                     | 85.45                                                                  |
| 4      | 1024         | 32            | 25.34                                                                      | 22.83                                                                     | 47.55                                                                  |
| 4      | 1024         | 64            | 22.08                                                                      | 19.51                                                                     | 43.78                                                                  |
| 4      | 1024         | 128           | 20.34                                                                      | 17.79                                                                     | 43.98                                                                  |
| 4      | 1024         | 256           | 19.67                                                                      | 16.94                                                                     | 46.83                                                                  |
| 4      | 1024         | 512           | 19.61                                                                      | 16.59                                                                     | 48.05                                                                  |
| 4      | 2048         | 16            | 49.47                                                                      | 43.94                                                                     | 126.89                                                                 |
| 4      | 2048         | 32            | 34.46                                                                      | 30.36                                                                     | 77.12                                                                  |
| 4      | 2048         | 64            | 28.09                                                                      | 23.39                                                                     | 76.69                                                                  |
| 4      | 2048         | 128           | 25.18                                                                      | 20.29                                                                     | 77.03                                                                  |
| 4      | 2048         | 256           | 23.53                                                                      | 18.62                                                                     | 76.90                                                                  |
| 4      | 2048         | 512           | 23.10                                                                      | 18.13                                                                     | 79.94                                                                  |

In summary, the average speedup from 2GPUs to 4GPUs is 21% (average), 23% (P50) and 11% (P90).

This was honestly surprising to me because multi-GPU training often scales sub-linearly because of the communication overhead. Frankly speaking, my understanding in multi-gpu inference is nescient, so I am wondering how this speedup is introduced from the compute perspective. Is this because the attention computation is being parallelized? I would much appreciate any education/thoughts/insights! Thanks again!

### Information

- [X] Docker
- [ ] The CLI directly

### Tasks

- [X] An officially supported command
- [ ] My own modifications

### Reproduction

N/A - this post is more of a theoretical question from my empirical findings.

### Expected behavior

N/A

**Comment by linkedlist771** [comment]
![image](https://github.com/huggingface/text-generation-inference/assets/72634327/fc73dff1-95d6-4901-82a8-294f1f8c7cc9)
Hi, I used four A100 to do the same benchmark as you with this [https://github.com/vllm-project/vllm/blob/main/benchmarks/benchmark_throughput.py](https://github.com/vllm-project/vllm/blob/main/benchmarks/benchmark_throughput.py). I am doing the async request for about 100 prompts, only to find that multi-GPUs do not imporve its performance greatl?(By the way, I increase the requests speed gradually and the value are the convergent value.)

**Comment by Rogerwyf** [comment]
> By the way, I increase the requests speed gradually and the value are the convergent value.

@linkedlist771 I'm curious what you mean by "increase the requests speed gradually?" since you're sending the 100 prompts async? 

My next step is to run the benchmarks on A100s and will report back!

**Comment by linkedlist771** [comment]
Hi, this is related with the bechmark test scripts in the `vllm` project. You can check the [code here](https://github.com/vllm-project/vllm/blob/main/benchmarks/benchmark_serving.py), in this fucntion:
```python
async def get_request(
    input_requests: List[Tuple[str, int, int]],
    request_rate: float,
) -> AsyncGenerator[Tuple[str, int, int], None]:
    input_requests = iter(input_requests)
    for request in input_requests:
        yield request

        if request_rate == float("inf"):
            # If the request rate is infinity, then we don't need to wait.
            continue
        # Sample the request interval from the exponential distribution.
        interval = np.random.exponential(1.0 / request_rate)
        # The next request will be sent after the interval.
        await asyncio.sleep(interval)
```

The request sending speed is decided by the exponential distribution, so I can gradually `increase the requests speed` by make the request interval time shorter.  

So nice to here that you want share the results.

**Comment by Rogerwyf** [comment]
Ah I see - so there could possibly be **concurrent** requests. Have you tried benchmarking by sending one request only after another being finished for a more rigorous result? My question is more on the request-level so I wanted to isolate the impact of concurrency from this benchmark.

**Comment by linkedlist771** [comment]
Sorry about that, I didn't test the `synchronous` cases. My use situation only requires the `asynchronous` performance. If you could test the `synchronous` cases , I would be happy to know that.

**Comment by Rogerwyf** [comment]
I was doing async cases too actually - the model is processing only one request a time.

Btw, it seems that Databricks has done some benchmarks and observed the similar results (I guess this does not apply specifically to TGI only): See Figure 4 in https://www.databricks.com/blog/llm-inference-performance-engineering-best-practices, and I guess their explanation sort of answered my original question!

**Comment by linkedlist771** [comment]
Very good blog, recently I have try to improve its' qps by the `round-robin` algorithm which greatly improve its qps by `58%` compared with distributed deployment.  Hopefully, I would make a `Pull Request` if I have modified my code.

**Comment by OlivierDehaene** [comment]
Hello!
Yes, this behaviour is expected. This makes selecting the best target for each model very tricky as you need to balance the latency with sharding, the VRAM for the best throughput and the economics. 
At HF we often pick a bigger instance to serve the models through the API because of this.

Note that their is even a bigger speedup if you have NVLINK between the GPUs, which as far as I know, is not the case for g5 instances.

**Comment by muhammad-asn** [comment]
Very interesting, do you plan using A100 with single GPU or still multiple-gpu? @Rogerwyf

**Comment by Rogerwyf** [comment]
@muhammad-asn For some of the big models (e.g, 34b, 70b), single GPU is simply not an option, so my the purpose of my benchmark here is to see the tradeoff between having one big instance over 8 GPUs or two instances with each over 4 GPUs.

Surprisingly, when I ran the same benchmark with `llama-2-70b-hf-chat` on `p4de.24xlarge` (4gpu vs 8 gpu), I observed some **performance slowdown (15% on average)** when model is sharded over multiple GPUs and I've verified that the instance has nVlink. - Do you might have some insights on this? @OlivierDehaene

My next step is to run benchmark on `p4d.24xlarge` and see if the same thing happens there.
| Model           | Instance      | \# GPU | Input Length | Output Limit | Average  | Median | P99  |
| --------------- | ------------- | ------ | ------------ | ------------------- | -------------------------------------------------------------------------- | ------------------------------------------------------------------------- | ---------------------------------------------------------------------- |
| llama2-70b-chat | p4de.24xlarge | 4      | 256          | 16                  | 44.93                                                                      | 45.18                                                                     | 50.30                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 256          | 32                  | 40.51                                                                      | 39.10                                                                     | 47.01                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 256          | 64                  | 39.35                                                                      | 39.21                                                                     | 41.54                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 256          | 128                 | 40.17                                                                      | 40.19                                                                     | 41.42                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 256          | 256                 | 40.21                                                                      | 40.24                                                                     | 41.53                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 256          | 512                 | 40.53                                                                      | 40.01                                                                     | 46.04                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 512          | 16                  | 46.38                                                                      | 41.73                                                                     | 69.54                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 512          | 32                  | 41.06                                                                      | 40.92                                                                     | 45.95                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 512          | 64                  | 40.56                                                                      | 40.42                                                                     | 43.60                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 512          | 128                 | 42.67                                                                      | 40.81                                                                     | 47.84                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 512          | 256                 | 40.68                                                                      | 39.86                                                                     | 46.82                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 512          | 512                 | 40.08                                                                      | 39.99                                                                     | 42.07                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 1024         | 16                  | 47.84                                                                      | 46.73                                                                     | 64.94                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 1024         | 32                  | 44.00                                                                      | 43.43                                                                     | 51.31                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 1024         | 64                  | 46.12                                                                      | 47.39                                                                     | 53.12                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 1024         | 128                 | 41.44                                                                      | 41.23                                                                     | 45.41                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 1024         | 256                 | 41.39                                                                      | 41.20                                                                     | 44.47                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 1024         | 512                 | 41.85                                                                      | 41.40                                                                     | 46.83                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 2048         | 16                  | 63.46                                                                      | 60.88                                                                     | 117.13                                                                 |
| llama2-70b-chat | p4de.24xlarge | 4      | 2048         | 32                  | 54.05                                                                      | 52.67                                                                     | 65.98                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 2048         | 64                  | 49.74                                                                      | 48.97                                                                     | 58.89                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 2048         | 128                 | 48.76                                                                      | 48.41                                                                     | 53.86                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 2048         | 256                 | 47.46                                                                      | 47.18                                                                     | 52.20                                                                  |
| llama2-70b-chat | p4de.24xlarge | 4      | 2048         | 512                 | 47.86                                                                      | 47.65                                                                     | 50.46                                                                  |
| llama2-70b-chat | p4de.24xlarge | 8      | 256          | 16                  | 69.13                                                                      | 64.18                                                                     | 109.15                                                                 |
| llama2-70b-chat | p4de.24xlarge | 8      | 256          | 32                  | 54.38                                                                      | 54.12                                                                     | 65.91                                                                  |
| llama2-70b-chat | p4de.24xlarge | 8      | 256          | 64                  | 50.94                                                                      | 49.09                                                                     | 71.34                                                                  |
| llama2-70b-chat | p4de.24xlarge | 8      | 256          | 128                 | 46.41                                                                      | 44.55                                                                     | 57.45                                                                  |
| llama2-70b-chat | p4de.24xlarge | 8      | 256          | 256                 | 44.33                                                                      | 42.65                                                                     | 53.12                                                                  |
| llama2-70b-chat | p4de.24xlarge | 8      | 256          | 512                 | 43.96                                                                      | 42.65                                                                     | 62.14                                                                  |
| llama2-70b-chat | p4de.24xlarge | 8      | 512          | 16                  | 67.18                                                                      | 65.23                                                                     | 100.44                                                                 |
| llama2-70b-chat | p4de.24xlarge | 8      | 512          | 32                  | 54.76                                                                      | 52.91                                                                     | 84.34                                                                  |
| llama2-70b-chat | p4de.24xlarge | 8      | 512          | 64                  | 50.36                                                                      | 47.10                                                                     | 131.86                                                                 |
| llama2-70b-chat | p4de.24xlarge | 8      | 512          | 128                 | 49.87                                                                      | 44.95                                                                     | 131.15                                                                 |
| llama2-70b-chat | p4de.24xlarge | 8      | 512          | 256                 | 45.02                                                                      | 43.08                                                                     | 75.67                                                                  |
| llama2-70b-chat | p4de.24xlarge | 8      | 512          | 512                 | 44.55                                                                      | 42.85                                                                     | 65.01                                                                  |
| llama2-70b-chat | p4de.24xlarge | 8      | 1024         | 16                  | 70.67                                                                      | 66.14                                                                     | 119.72                                                                 |
| llama2-70b-chat | p4de.24xlarge | 8      | 1024         | 32                  | 63.78                                                                      | 54.08                                                                     | 267.14                                                                 |
| llama2-70b-chat | p4de.24xlarge | 8      | 1024         | 64                  | 53.45                                                                      | 49.91                                                                     | 166.67                                                                 |
| llama2-70b-chat | p4de.24xlarge | 8      | 1024         | 128                 | 46.61                                                                      | 45.60                                                                     | 59.67                                                                  |
| llama2-70b-chat | p4de.24xlarge | 8      | 1024         | 256                 | 46.29                                                                      | 43.63                                                                     | 90.41                                                                  |
| llama2-70b-chat | p4de.24xlarge | 8      | 1024         | 512                 | 44.26                                                                      | 42.84                                                                     | 67.44                                                                  |
| llama2-70b-chat | p4de.24xlarge | 8      | 2048         | 16                  | 77.15                                                                      | 69.71                                                                     | 204.49                                                                 |
| llama2-70b-chat | p4de.24xlarge | 8      | 2048         | 32                  | 60.24                                                                      | 58.19                                                                     | 87.49                                                                  |
| llama2-70b-chat | p4de.24xlarge | 8      | 2048         | 64                  | 54.33                                                                      | 49.14                                                                     | 84.91                                                                  |
| llama2-70b-chat | p4de.24xlarge | 8      | 2048         | 128                 | 50.48                                                                      | 47.07                                                                     | 122.97                                                                 |
| llama2-70b-chat | p4de.24xlarge | 8      | 2048         | 256                 | 45.31                                                                      | 44.39                                                                     | 51.61                                                                  |
| llama2-70b-chat | p4de.24xlarge | 8      | 2048         | 512                 | 44.82                                                                      | 43.31                                                                     | 56.17                                                                  |

**Comment by robmsmt** [comment]
Very interesting, thank you for the results. Are you able to provide all the parameters that you are using to start TGI?

I am attempting something similar on GCP's GKE (using the 40GB A100's and am having issues, it only seems to run with --disable-custom-kernels ) as well as only using 1 GPU of the 4 despite sharding and num_shards being on their defaults.

**Comment by Rogerwyf** [comment]
@robmsmt 
> Are you able to provide all the parameters that you are using to start TGI?

Yes, I'm using TGI v1.1.0 with most default arguments except for max input/total tokens

```
- '--num-shard'
- '8' # this depends on the number of GPUs
- '--max-input-length'
- '2048'
- '--max-total-tokens'
- '2560'
```

> I am attempting something similar on GCP's GKE (using the 40GB A100's and am having issues, it only seems to run with --disable-custom-kernels )

Maybe it's worth checking out the CUDA version?

**Comment by muhammad-asn** [comment]
Not sure but it might related to the latency of the multiple GPU. The bigger memory will loaded more the copies of the model partition hence it will increase latency on A100 80GB. Cmiiw @Rogerwyf
